test/jdk/java/util/zip/CopyZipFile.java

```diff
@@ -0,0 +1,204 @@
+        OutputStream os = new ByteArrayOutputStream();
+        ZipOutputStream zos = new ZipOutputStream(os);
+        while((entry = zis.getNextEntry())!=null) {
+            // ZipInputStream.getNextEntry() only reads the Local File Header of a zip entry,
+            // so for the zip file we've just generated the ZipEntry fields 'size', 'compressedSize`
+            // and 'crc' for deflated entries should be uninitialized (i.e. '-1').
+            System.out.println(
+                String.format("name=%s, clen=%d, len=%d, crc=%d",
+                              entry.getName(), entry.getCompressedSize(), entry.getSize(), entry.getCrc()));
+            if (entry.getMethod() == ZipEntry.DEFLATED &&
+                (entry.getCompressedSize() != -1 || entry.getSize() != -1 || entry.getCrc() != -1)) {
+                throw new Exception("'size', 'compressedSize' and 'crc' shouldn't be initialized at this point.");
+            }
+            zos.putNextEntry(entry);
+            zis.transferTo(zos);
+            // After all the data belonging to a zip entry has been inflated (i.e. after ZipInputStream.read()
+            // returned '-1'), it is guaranteed that the ZipInputStream will also have consumed the Data
+            // Descriptor (if any) after the data and will have updated the 'size', 'compressedSize' and 'crc'
+            // fields of the ZipEntry object.
+            System.out.println(
+                String.format("name=%s, clen=%d, len=%d, crc=%d\n",
+                              entry.getName(), entry.getCompressedSize(), entry.getSize(), entry.getCrc()));
+            if (entry.getCompressedSize() == -1 || entry.getSize() == -1) {
+                throw new Exception("'size' and 'compressedSize' must be initialized at this point.");
+            }
+        }
+        zos.close();
+
+        // Now we read all the entries of the initially generated zip file with the help
+        // of the ZipFile class. The ZipFile class reads all the zip entries from the Central
+        // Directory which must have accurate information for size, compressedSize and crc.
+        // This means that all ZipEntry objects returned from ZipFile will have correct
+        // settings for these fields.
+        // I the compression level was different in the initial zip file (which we can't find
+        // out any more now because the zip file format doesn't record this information) the
+        // size of the re-compressed entry we are writing to the ZipOutputStream might differ
+        // from the original compressed size recorded in the ZipEntry. This would result in an
+        // "invalid entry compressed size" ZipException if ZipOutputStream wouldn't ignore
+        // the implicitely set compressed size attribute of ZipEntries read from a ZipFile
+        // or ZipInputStream.
+        os = new ByteArrayOutputStream();
+        zos = new ZipOutputStream(os);
+        ZipFile zf = new ZipFile(ZIP_FILE);
```

Comment:
Any reason to not use try with resources here as well (and below)

Reply:
Done